CRM/MembershipExtras/Hook/Pre/MembershipPeriodUpdater.php

```diff
@@ -0,0 +1,322 @@
+   * membership edit was triggered by completing
+   * a contribution.
+   */
+  private $contribution = NULL;
+
+  /**
+   * The period membership details
+   */
+  private $membership;
+
+  /**
+   * The membership new join date
+   */
+  private $calculatedMembershipJoinDate;
+
+  /**
+   * The membership new end date
+   */
+  private $calculatedMembershipEndDate;
+
+  private $firstActivatedPeriod;
+
+  private $lastActivatedPeriod;
+
+  private $firstActivatedPeriodStartDate;
+
+  private $firstActivatedPeriodEndDate;
+
+  private $lastActivatedPeriodStartDate;
+
+  private $lastActivatedPeriodEndDate;
+
+  public function __construct($membershipId, &$membershipHookParams, $contributionID) {
+    $this->membershipHookParams = &$membershipHookParams;
+    $this->setMembership($membershipId);
+    $this->setContribution($contributionID);
+    $this->setCalculatedMembershipJoinDate();
+    $this->setCalculatedMembershipEndDate();
+  }
+
+  private function setMembership($membershipId) {
+    $this->membership =  civicrm_api3('Membership', 'get', [
+      'id' => $membershipId,
+      'return' => ['membership_type_id.name', 'join_date', 'end_date'],
+      'sequential' => 1,
+    ])['values'][0];
+  }
+
+  private function setContribution($contributionID) {
+    if (empty($contributionID)) {
+      return;
+    }
+
+    $paymentContribution = civicrm_api3('Contribution', 'get', [
+      'sequential' => 1,
+      'id' => $contributionID,
+      'return' => ['id', 'contribution_recur_id', 'contribution_status_id'],
+    ]);
+
+    $contribution =  NULL;
+    if (!empty($paymentContribution['values'][0])) {
+      $contribution = $paymentContribution['values'][0];
+    }
+
+    $this->contribution = $contribution;
+  }
+
+  private function setCalculatedMembershipJoinDate() {
+    if (empty($this->membershipHookParams['join_date'])) {
+      $membershipJoinDate = date('Ymd', strtotime($this->membership['join_date']));
+    } else {
+      $membershipJoinDate = date('Ymd', strtotime($this->membershipHookParams['join_date']));
+    }
+
+    $this->calculatedMembershipJoinDate = $membershipJoinDate;
+  }
+
+  private function setCalculatedMembershipEndDate() {
+    $membershipEndDate = NULL;
+    if (!empty($this->membershipHookParams['end_date'])) {
+      $membershipEndDate = date('Ymd', strtotime($this->membershipHookParams['end_date']));
+    }
+    // on membership 'bulk update' the field name is membership_end_date and not end_date
+    elseif(!empty($this->membershipHookParams['membership_end_date'])) {
+      $membershipEndDate = date('Ymd', strtotime($this->membershipHookParams['membership_end_date']));
+    }
+    elseif (!empty($this->membership['end_date'])){
+      $membershipEndDate = date('Ymd', strtotime($this->membership['end_date']));
+    }
+
+    $this->calculatedMembershipEndDate = $membershipEndDate;
+  }
+
+  public function process() {
+    $this->rectifyPendingPeriodUponPaymentCompletion();
```

Comment:
Maybe `activatePendingPeriodUponPaymentCompletion()` is clearer?

Reply:
it does not only activate the period but also might update its date in some cases as mentioned in rectifyPendingPeriodUponPaymentCompletion() method description, that's why i think rectify is a better term here.